### PR TITLE
Add ernie4 5 moe

### PR DESCRIPTION
**Pull request by cheng221**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR changes

主要改动在于：
1. Ernie 4.5 Model
2. 新增通用模块 moe、pipeline model 
3. model.generate使用的部分接口通用
### Description

1. PR 基于通用模块，进行了Ernie4.5 Moe组网代码精简。支持PT权重，EP训练推理
2. 新增MTP Loss 跟 通用Pipeline Model（自行验证过）
3. 将update_model_kwargs_for_generation、prepare_inputs_for_generation 移动到PretrainedModel

遗留TODO:
1. 新增模块的单测，正在补充ernie 4.5 moe 模型单测 （这个PR完成）
2. Tensor parallel mapping 移到父类实现（下一个PR完成）

### CODE REVIEW

File: paddleformers/transformers/ernie4_5/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# limitations under the License.
+
+import transformers as hf
+
+from ..tokenizer_utils import warp_tokenizer
+
+__all__ = ["Ernie4_5Tokenizer"]
+
+
+Ernie4_5Tokenizer = warp_tokenizer(hf.LlamaTokenizer)
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5/tokenizer.py:22]: 保留Ernie4_5Tokenizer原因？

File: paddleformers/transformers/ernie4_5/modeling.py [hunk 2]
```diff
@@ -171,7 +171,9 @@ def __init__(self, config, layer_idx=0):
                 ), f"num_heads: {self.num_key_value_heads}, tensor_parallel_degree: {config.tensor_parallel_degree}"
                 self.num_key_value_heads = self.num_key_value_heads // config.tensor_parallel_degree
         if self.is_gqa:
-            logger.info(f"use GQA - num_heads: {self.num_heads}- num_key_value_heads: {self.num_key_value_heads}")
+            logger.warning_once(
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5/modeling.py]: self.is_gqa：本来就是gqa模型是不是不需要额外判断是否是gqa，可以把self.is_gqa相关逻辑都删除

File: paddleformers/transformers/ernie4_5/modeling.py [hunk 3]
```diff
@@ -452,47 +440,53 @@ def _get_tensor_parallel_mappings(cls, config, is_split=True):
...
-            final_actions = {}
-
-            base_actions = {
-                # Column Linear
-                "layers.0.self_attn.q_proj.weight": partial(fn, is_column=True),
-                "layers.0.self_attn.k_proj.weight": partial(fn, is_column=True),
-                "layers.0.self_attn.v_proj.weight": partial(fn, is_column=True),
-                "layers.0.mlp.gate_proj.weight": partial(fn, is_column=True),
-                "layers.0.mlp.up_proj.weight": partial(fn, is_column=True),
+        LAYER_COLWISE = [
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5/modeling.py:419]: 是否验证过与ERNIEKit中模型的输出logit是否一致

File: paddleformers/transformers/ernie4_5_moe/configuration.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        hidden_act="silu",
+        initializer_range=0.02,
+        rms_norm_eps=1e-6,
+        use_cache=False,
+        use_flash_attention=True,
+        use_rmsnorm=True,
+        pad_token_id=0,
+        bos_token_id=1,
+        eos_token_id=2,
+        fuse_swiglu=False,
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/configuration.py]: 这个是不是可以放在PretrainedConfig中

File: paddleformers/transformers/ernie4_5_moe/configuration.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        use_bias=False,
+        rope_theta=10000,
+        max_sequence_length=None,
+        ignored_index=-100,
+        attention_dropout_prob=0.0,
+        hidden_dropout_prob=0.0,
+        compression_ratio: float = 1.0,
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/configuration.py]: compression_ratio没用的话可以去掉

File: paddleformers/transformers/ernie4_5_moe/configuration.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        moe_norm_min=1e-12,
+        moe_aux_loss_lambda=1e-2,
+        moe_z_loss_lambda=1e-4,
+        moe_orthogonal_loss_lambda=1e-2,
+        sinkhorn_2gate=True,
+        sinkhorn_temp=3e-2,
+        global_aux_loss=False,
+        moe_dropout_prob=0.0,
+        moe_group="dummy",
+        moe_gate="topk",
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/configuration.py]: moe_gate这个需要吗

File: paddleformers/transformers/ernie4_5_moe/configuration.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        moe_all_to_all_dropout: float = 0.0,
+        moe_k=2,
+        moe_use_aux_free: bool = True,
+        moe_group_experts: bool = False,
+        moe_group_orthogonal_loss: bool = True,
+        enable_delay_scale_loss: bool = True,
+        num_acc_steps: int = 1,
+        fuse_gate_detach_matmul: bool = False,
+        moe_use_hard_gate=False,
+        moe_dense_experts_token_type_id=3,
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/configuration.py]: token_type_ids纯文是否会用到

File: paddleformers/transformers/ernie4_5_moe/configuration.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        self.moe_use_aux_free = moe_use_aux_free
+        self.fuse_gate_detach_matmul = fuse_gate_detach_matmul
+        self.moe_use_hard_gate = moe_use_hard_gate
+        self.moe_dense_experts_token_type_id = moe_dense_experts_token_type_id
+        self.num_nextn_predict_layers = num_nextn_predict_layers
+        self.multi_token_pred_lambda = multi_token_pred_lambda
+        self.enable_mtp_magic_send = enable_mtp_magic_send
+        self.use_recompute_mtp = use_recompute_mtp
+
+        self.register_unsavable_keys(
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/configuration.py:237]: unsavable_keys是否覆盖全面，保证存下的config key和加载的config key数量一致

File: paddleformers/transformers/ernie4_5_moe/modeling.py [hunk 9]
```diff
@@ -0,0 +1,1131 @@
...
+        Returns:
+            Union: Various output combinations depending on arguments:
+                - Base case: Hidden states tensor
+                - With attention: Tuple of (hidden_states, attention_weights)
+                - With cache: Tuple of (hidden_states, cached_key_value)
+                - With MoE: May include gate logits in output tuple
+        """
+        residual = hidden_states
+
+        if token_type_ids is not None:
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/modeling.py]: 纯文模型中 token_type_ids可以去除

File: paddleformers/transformers/ernie4_5_moe/modeling.py [hunk 10]
```diff
@@ -0,0 +1,1131 @@
...
+
+        if self.config.num_nextn_predict_layers > 0:
+            self.mtp_block = paddle.nn.LayerList(
+                [
+                    Ernie4_5_MoeDecoderLayer(config, layer_idx)
+                    for layer_idx in range(self.config.num_nextn_predict_layers)
+                ]
+            )
+
+            GeneralNorm.create(
```
Comment (lugimzzz) [inline paddleformers/transformers/ernie4_5_moe/modeling.py]: 这个是不是多写了

File: paddleformers/transformers/ernie4_5/tokenizer.py (hunk 1 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5/tokenizer.py:22]: 上周说不是要在wrap一下嘛，这个不用了就删掉

File: paddleformers/transformers/ernie4_5/modeling.py (hunk 2 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5/modeling.py]: done

File: paddleformers/transformers/ernie4_5_moe/configuration.py (hunk 4 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/configuration.py]: done

File: paddleformers/transformers/ernie4_5_moe/configuration.py (hunk 5 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/configuration.py]: done

File: paddleformers/transformers/ernie4_5_moe/configuration.py (hunk 8 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/configuration.py:237]: ernie4_5 dense and moe done

File: paddleformers/transformers/ernie4_5_moe/modeling.py (hunk 9 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/modeling.py]: done

File: paddleformers/transformers/ernie4_5_moe/modeling.py (hunk 10 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/modeling.py]: done

File: paddleformers/transformers/ernie4_5_moe/configuration.py (hunk 6 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5_moe/configuration.py]: 看了一下只有nn.moe的log有用到，可以删去

File: paddleformers/transformers/ernie4_5/modeling.py (hunk 3 above)
Comment (cheng221) [reply paddleformers/transformers/ernie4_5/modeling.py:419]: pp2，tp4训练配置下。目前lmhead输出的logits差距平均0.15，正在排查原因

**Review by lugimzzz — APPROVED** [review]
lgtm
